lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -533,6 +645,13 @@ LogicalResult XorOp::canonicalize(XorOp op, PatternRewriter &rewriter) {
   if (tryFlatteningOperands(op, rewriter))
     return success();
 
+  // xor(a, b) -> concat(0, xor(a[n-1:0], b[n-1:0])), where n is the widest
+  // demanded width
+  if (narrowOperationWidth(op, inputs, /* narrowTrailingBits= */ true,
```

Comment:
I'd recommend putting these checks into the ExtractOp canonicalization hook.  It will be more efficient because it only triggers on extracts, instead of having to analyze the use-list of every arithmetic operation.

Reply:
would that be okay in the general case? Say

```mlir
%a = add %arg0, %arg1 : i10
%b = extract %a from 0 : i4
%c = extract %a from 1 : i5
```

If i do this canonicalization on extract, i'd have to rewrite `%a` when canonicalizing `%b`. does MLIR allow that?